powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/converter/v3/RecoveryPrivateKeyConverter.java

```diff
@@ -0,0 +1,233 @@
+        String masterDbEncryptionKeyBase64 = powerAuthServiceConfiguration.getMasterDbEncryptionKey();
+
+        // In case master DB encryption key does not exist, do not encrypt the server private key
+        if (masterDbEncryptionKeyBase64 == null || masterDbEncryptionKeyBase64.isEmpty()) {
+            return new RecoveryPrivateKey(EncryptionMode.NO_ENCRYPTION, BaseEncoding.base64().encode(recoveryPrivateKey));
+        }
+
+        try {
+            // Convert master DB encryption key
+            SecretKey masterDbEncryptionKey = keyConvertor.convertBytesToSharedSecretKey(BaseEncoding.base64().decode(masterDbEncryptionKeyBase64));
+
+            // Derive secret key from master DB encryption key and application ID
+            SecretKey secretKey = deriveSecretKey(masterDbEncryptionKey, applicationId);
+
+            // Generate random IV
+            byte[] iv = keyGenerator.generateRandomBytes(16);
+
+            // Encrypt recoveryPrivateKey using secretKey with generated IV
+            byte[] encrypted = aesEncryptionUtils.encrypt(recoveryPrivateKey, iv, secretKey);
+
+            // Generate output bytes as encrypted + IV
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            baos.write(iv);
+            baos.write(encrypted);
+            byte[] record = baos.toByteArray();
+
+            // Base64-encode output and create ServerPrivateKey instance
+            String encryptedKeyBase64 = BaseEncoding.base64().encode(record);
+
+            // Return encrypted record including encryption mode
+            return new RecoveryPrivateKey(EncryptionMode.AES_HMAC, encryptedKeyBase64);
+
+        } catch (InvalidKeyException ex) {
+            logger.error(ex.getMessage(), ex);
+            // Rollback is not required, cryptography methods are executed before database is used for writing
+            throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_KEY_FORMAT);
+        } catch (GenericCryptoException ex) {
+            logger.error(ex.getMessage(), ex);
+            // Rollback is not required, cryptography methods are executed before database is used for writing
+            throw localizationProvider.buildExceptionForCode(ServiceError.GENERIC_CRYPTOGRAPHY_ERROR);
+        } catch (CryptoProviderException ex) {
+            logger.error(ex.getMessage(), ex);
+            // Rollback is not required, cryptography methods are executed before database is used for writing
+            throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_CRYPTO_PROVIDER);
+        } catch (IOException ex) {
+            logger.error(ex.getMessage(), ex);
+            // Rollback is not required, serialization is executed before database is used for writing
+            throw localizationProvider.buildExceptionForCode(ServiceError.ENCRYPTION_FAILED);
+        }
+    }
+
+    /**
+     * Derive secret key from master DB encryption key, user ID and activation ID.<br/>
```

Comment:
There's no user ID and activation ID involved in this derivation. Looks like copy-paste from similar derivation code.

Reply:
Thanks for noticing, I will fix it.